File: codes/dart/array_some.dart
import "dart:convert";

void main() {
    dynamic jsonStringify(dynamic anything, { bool pretty = false, String indent = "    " }) {
		if (pretty == true) {
			final jsonEncoder = new JsonEncoder.withIndent(indent);
			return jsonEncoder.convert(anything);
		}
		final jsonEncoder = new JsonEncoder();
		return jsonEncoder.convert(anything).split(',').join(', ');
    }

    dynamic arraySomeV1(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.some() function
        dynamic isConditionMatch = false;
        for (final objectEntry in anArray.asMap().entries) {
            final arrayItemIndex = objectEntry.key;
            final arrayItem = objectEntry.value;
            isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == true) break;
        }
        return isConditionMatch;
    }

    dynamic arraySomeV2(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.some() function
        dynamic isConditionMatch = false;
        for (final objectEntry in anArray.asMap().entries) {
            final arrayItemIndex = objectEntry.key;
            final arrayItem = objectEntry.value;
            isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == true) return isConditionMatch;
        }
        return isConditionMatch;
    }

    dynamic arraySomeV3(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.some() function
        for (final objectEntry in anArray.asMap().entries) {
            final arrayItemIndex = objectEntry.key;
            final arrayItem = objectEntry.value;
            dynamic isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == true) return true;
        }
        return false;
    }

    dynamic arraySomeV4(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.some() function
        for (final objectEntry in anArray.asMap().entries) {
            final arrayItemIndex = objectEntry.key;
            final arrayItem = objectEntry.value;
            if (callbackFunction(arrayItem, arrayItemIndex, anArray) == true) return true;
        }
        return false;
    }

    print("\n// JavaScript-like Array.some() in Dart List");

    final numbers = <dynamic>[12, 34, 27, 23, 65, 93, 36, 87, 4, 254];
    print("numbers: ${jsonStringify(numbers)}");

    dynamic isAnyNumberLessThan500;
    dynamic isAnyNumberMoreThan500;

    print("// using JavaScript-like Array.some() function \"arraySomeV1\"");

    isAnyNumberLessThan500 = arraySomeV1((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is any number < 500: ${isAnyNumberLessThan500}");
    // is any number < 500: true

    isAnyNumberMoreThan500 = arraySomeV1((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is any number > 500: ${isAnyNumberMoreThan500}");
    // is any number > 500: false

    print("// using JavaScript-like Array.some() function \"arraySomeV2\"");

    isAnyNumberLessThan500 = arraySomeV2((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is any number < 500: ${isAnyNumberLessThan500}");
    // is any number < 500: true

    isAnyNumberMoreThan500 = arraySomeV2((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is any number > 500: ${isAnyNumberMoreThan500}");
    // is any number > 500: false

    print("// using JavaScript-like Array.some() function \"arraySomeV3\"");

    isAnyNumberLessThan500 = arraySomeV3((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is any number < 500: ${isAnyNumberLessThan500}");
    // is any number < 500: true

    isAnyNumberMoreThan500 = arraySomeV3((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is any number > 500: ${isAnyNumberMoreThan500}");
    // is any number > 500: false

    print("// using JavaScript-like Array.some() function \"arraySomeV4\"");

    isAnyNumberLessThan500 = arraySomeV4((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is any number < 500: ${isAnyNumberLessThan500}");
    // is any number < 500: true

    isAnyNumberMoreThan500 = arraySomeV4((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is any number > 500: ${isAnyNumberMoreThan500}");
    // is any number > 500: false

    print("// using Dart Array.some() built-in method \"Array.any()\"");

    isAnyNumberLessThan500 = numbers.any((dynamic number) => (number < 500));
    print("is any number < 500: ${isAnyNumberLessThan500}");
    // is any number < 500: true

    isAnyNumberMoreThan500 = numbers.any((dynamic number) => (number > 500));
    print("is any number > 500: ${isAnyNumberMoreThan500}");
    // is any number > 500: false

    print("\n// JavaScript-like Array.some() in Dart List of Maps");

    final products = <dynamic>[
        <String, dynamic>{
            "code": "pasta",
            "price": 321
        },
        <String, dynamic>{
            "code": "bubble_gum",
            "price": 233
        },
        <String, dynamic>{
            "code": "potato_chips",
            "price": 5
        },
        <String, dynamic>{
            "code": "towel",
            "price": 499
        }
    ];
    print("products: ${jsonStringify(products, pretty: true)}");

    dynamic isAnyProductPriceLessThan500;
    dynamic isAnyProductPriceMoreThan500;

    print("// using JavaScript-like Array.some() function \"arraySomeV1\"");

    isAnyProductPriceLessThan500 = arraySomeV1((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is any product price < 500: ${isAnyProductPriceLessThan500}");
    // is any product price < 500: true

    isAnyProductPriceMoreThan500 = arraySomeV1((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is any product price > 500: ${isAnyProductPriceMoreThan500}");
    // is any product price > 500: false

    print("// using JavaScript-like Array.some() function \"arraySomeV2\"");

    isAnyProductPriceLessThan500 = arraySomeV2((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is any product price < 500: ${isAnyProductPriceLessThan500}");
    // is any product price < 500: true

    isAnyProductPriceMoreThan500 = arraySomeV2((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is any product price > 500: ${isAnyProductPriceMoreThan500}");
    // is any product price > 500: false

    print("// using JavaScript-like Array.some() function \"arraySomeV3\"");

    isAnyProductPriceLessThan500 = arraySomeV3((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is any product price < 500: ${isAnyProductPriceLessThan500}");
    // is any product price < 500: true

    isAnyProductPriceMoreThan500 = arraySomeV3((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is any product price > 500: ${isAnyProductPriceMoreThan500}");
    // is any product price > 500: false

    print("// using JavaScript-like Array.some() function \"arraySomeV4\"");

    isAnyProductPriceLessThan500 = arraySomeV4((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is any product price < 500: ${isAnyProductPriceLessThan500}");
    // is any product price < 500: true

    isAnyProductPriceMoreThan500 = arraySomeV4((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is any product price > 500: ${isAnyProductPriceMoreThan500}");
    // is any product price > 500: false

    print("// using Dart Array.some() built-in method \"Array.any()\"");

    isAnyProductPriceLessThan500 = products.any((dynamic product) => (product["price"] < 500));
    print("is any product price < 500: ${isAnyProductPriceLessThan500}");
    // is any product price < 500: true

    isAnyProductPriceMoreThan500 = products.any((dynamic product) => (product["price"] > 500));
    print("is any product price > 500: ${isAnyProductPriceMoreThan500}");
    // is any product price > 500: false
}

File: codes/csharp/array_some.cs
using System; // Console, Func<>, Action<>, Action
using System.Linq; // Enumerable
using System.Collections.Generic; // Dictionary<>, List<>, KeyValuePair<>, IEnumerable<>

class Program {
    static void Main(string[] Args) {
        dynamic IsNumeric = (Func<dynamic, bool>)((dynamic Anything) => (Anything is sbyte || Anything is byte || Anything is short || Anything is ushort || Anything is int || Anything is uint || Anything is long || Anything is ulong || Anything is float || Anything is double || Anything is decimal));

        string JsonStringify(dynamic Anything, bool Pretty = false, string Indent = "    ") {
            int IndentLevel = 0;
            string JsonStringifyInner(dynamic AnythingInner, string IndentInner) {
                if (AnythingInner == null) return "null";
                if (AnythingInner is string) return "\"" + (string)AnythingInner + "\"";
                if ((IsNumeric(AnythingInner) == true) || AnythingInner is bool) return AnythingInner.ToString().Replace(",", ".");
                if (AnythingInner is List<dynamic>) {
                    IndentLevel += 1;
                    string Result = ((Pretty == true) ? ("[" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : "[");
                    int ArrayItemIndex = 0;
                    foreach (dynamic ArrayItem in (List<dynamic>)AnythingInner) {
                        Result += JsonStringifyInner(ArrayItem, IndentInner);
                        if ((ArrayItemIndex + 1) != ((List<dynamic>)AnythingInner).Count) Result += ((Pretty == true) ? ("," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : ", ");
                        ArrayItemIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += ((Pretty == true) ? (Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "]") : "]");
                    return Result;
                }
                if (AnythingInner is Dictionary<string, dynamic>) {
                    IndentLevel += 1;
                    string Result = ((Pretty == true) ? ("{" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : "{");
                    int ObjectIterationIndex = 0;
                    foreach (KeyValuePair<string, dynamic> ObjectEntry in (Dictionary<string, dynamic>)AnythingInner) {
                        string ObjectKey = ObjectEntry.Key;
                        dynamic ObjectValue = ObjectEntry.Value;
                        Result += "\"" + ObjectKey + "\": " + JsonStringifyInner(ObjectValue, IndentInner);
                        if ((ObjectIterationIndex + 1) != ((Dictionary<string, dynamic>)AnythingInner).Count) Result += ((Pretty == true) ? ("," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel))) : ", ");
                        ObjectIterationIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += ((Pretty == true) ? (Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "}") : "}");
                    return Result;
                }
                return "null";
            };
            return JsonStringifyInner(Anything, Indent);
        }

        bool ArraySomeV1(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.some() function
            bool IsConditionMatch = false;
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == true) break;
                ArrayItemIndex += 1;
            }
            return IsConditionMatch;
        }

        bool ArraySomeV2(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.some() function
            bool IsConditionMatch = false;
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == true) return IsConditionMatch;
                ArrayItemIndex += 1;
            }
            return IsConditionMatch;
        }

        bool ArraySomeV3(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.some() function
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                bool IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == true) return true;
                ArrayItemIndex += 1;
            }
            return false;
        }

        bool ArraySomeV4(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.some() function
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                if ((bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray) == true) return true;
                ArrayItemIndex += 1;
            }
            return false;
        }

        Console.WriteLine("\n// JavaScript-like Array.some() in C# List");

        dynamic Numbers = new List<dynamic>() {12, 34, 27, 23, 65, 93, 36, 87, 4, 254};
        Console.WriteLine($"Numbers: {JsonStringify(Numbers)}");

        dynamic IsAnyNumberLessThan500;
        dynamic IsAnyNumberMoreThan500;

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV1\"");

        IsAnyNumberLessThan500 = ArraySomeV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is any number < 500: {IsAnyNumberLessThan500}");
        // is any number < 500: True

        IsAnyNumberMoreThan500 = ArraySomeV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is any number > 500: {IsAnyNumberMoreThan500}");
        // is any number > 500: False

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV2\"");

        IsAnyNumberLessThan500 = ArraySomeV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is any number < 500: {IsAnyNumberLessThan500}");
        // is any number < 500: True

        IsAnyNumberMoreThan500 = ArraySomeV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is any number > 500: {IsAnyNumberMoreThan500}");
        // is any number > 500: False

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV3\"");

        IsAnyNumberLessThan500 = ArraySomeV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is any number < 500: {IsAnyNumberLessThan500}");
        // is any number < 500: True

        IsAnyNumberMoreThan500 = ArraySomeV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is any number > 500: {IsAnyNumberMoreThan500}");
        // is any number > 500: False

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV4\"");

        IsAnyNumberLessThan500 = ArraySomeV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is any number < 500: {IsAnyNumberLessThan500}");
        // is any number < 500: True

        IsAnyNumberMoreThan500 = ArraySomeV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is any number > 500: {IsAnyNumberMoreThan500}");
        // is any number > 500: False

        Console.WriteLine("// using C# Array.some() built-in method \"IEnumerable.Any()\"");

        IsAnyNumberLessThan500 = ((IEnumerable<dynamic>)Numbers).Any((Func<dynamic, bool>)((dynamic Number) => (Number < 500)));
        Console.WriteLine($"is any number < 500: {IsAnyNumberLessThan500}");
        // is any number < 500: True

        IsAnyNumberMoreThan500 = ((IEnumerable<dynamic>)Numbers).Any((Func<dynamic, bool>)((dynamic Number) => (Number > 500)));
        Console.WriteLine($"is any number > 500: {IsAnyNumberMoreThan500}");
        // is any number > 500: False

        Console.WriteLine("\n// JavaScript-like Array.some() in C# List of Dictionaries");

        dynamic Products = new List<dynamic>() {
            new Dictionary<string, dynamic>() {
                {"code", "pasta"},
                {"price", 321}
            },
            new Dictionary<string, dynamic>() {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            new Dictionary<string, dynamic>() {
                {"code", "potato_chips"},
                {"price", 5}
            },
            new Dictionary<string, dynamic>() {
                {"code", "towel"},
                {"price", 499}
            },
        };
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty: true)}");

        dynamic IsAnyProductPriceLessThan500;
        dynamic IsAnyProductPriceMoreThan500;

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV1\"");

        IsAnyProductPriceLessThan500 = ArraySomeV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is any product price < 500: {IsAnyProductPriceLessThan500}");
        // is any product price < 500: True

        IsAnyProductPriceMoreThan500 = ArraySomeV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is any product price > 500: {IsAnyProductPriceMoreThan500}");
        // is any product price > 500: False

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV2\"");

        IsAnyProductPriceLessThan500 = ArraySomeV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is any product price < 500: {IsAnyProductPriceLessThan500}");
        // is any product price < 500: True

        IsAnyProductPriceMoreThan500 = ArraySomeV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is any product price > 500: {IsAnyProductPriceMoreThan500}");
        // is any product price > 500: False

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV3\"");

        IsAnyProductPriceLessThan500 = ArraySomeV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is any product price < 500: {IsAnyProductPriceLessThan500}");
        // is any product price < 500: True

        IsAnyProductPriceMoreThan500 = ArraySomeV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is any product price > 500: {IsAnyProductPriceMoreThan500}");
        // is any product price > 500: False

        Console.WriteLine("// using JavaScript-like Array.some() function \"ArraySomeV4\"");

        IsAnyProductPriceLessThan500 = ArraySomeV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is any product price < 500: {IsAnyProductPriceLessThan500}");
        // is any product price < 500: True

        IsAnyProductPriceMoreThan500 = ArraySomeV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is any product price > 500: {IsAnyProductPriceMoreThan500}");
        // is any product price > 500: False

        Console.WriteLine("// using C# Array.some() built-in method \"IEnumerable.Any()\"");

        IsAnyProductPriceLessThan500 = ((IEnumerable<dynamic>)Products).Any((Func<dynamic, bool>)((dynamic Product) => (Product["price"] < 500)));
        Console.WriteLine($"is any product price < 500: {IsAnyProductPriceLessThan500}");
        // is any product price < 500: True

        IsAnyProductPriceMoreThan500 = ((IEnumerable<dynamic>)Products).Any((Func<dynamic, bool>)((dynamic Product) => (Product["price"] > 500)));
        Console.WriteLine($"is any product price > 500: {IsAnyProductPriceMoreThan500}");
        // is any product price > 500: False
    }
}
